File: src/main/java/com/dh/ondot/schedule/domain/vo/Sound.java
package com.dh.ondot.schedule.domain.vo;

import com.dh.ondot.schedule.domain.enums.RingTone;
import com.dh.ondot.schedule.domain.enums.SoundCategory;
import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Embeddable
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class Sound {
    @Enumerated(EnumType.STRING)
    @Column(name = "sound_category", nullable = false)
    private SoundCategory soundCategory;

    @Enumerated(EnumType.STRING)
    @Column(name = "ring_tone", nullable = false)
    private RingTone ringTone;

    @Column(name = "volume", nullable = false)
    private Double volume;

    public static Sound of(String soundCategory, String ringTone, Double volume) {
        return new Sound(SoundCategory.from(soundCategory), RingTone.from(ringTone), volume);
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/domain/vo/Sound.kt
package com.dh.ondot.schedule.domain.vo

import com.dh.ondot.schedule.domain.enums.RingTone
import com.dh.ondot.schedule.domain.enums.SoundCategory
import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated

@Embeddable
class Sound protected constructor() {

    @Enumerated(EnumType.STRING)
    @Column(name = "sound_category", nullable = false)
    var soundCategory: SoundCategory = SoundCategory.BRIGHT_ENERGY
        protected set

    @Enumerated(EnumType.STRING)
    @Column(name = "ring_tone", nullable = false)
    var ringTone: RingTone = RingTone.DANCING_IN_THE_STARDUST
        protected set

    @Column(name = "volume", nullable = false)
    var volume: Double = 0.5
        protected set

    private constructor(soundCategory: SoundCategory, ringTone: RingTone, volume: Double) : this() {
        this.soundCategory = soundCategory
        this.ringTone = ringTone
        this.volume = volume
    }

    companion object {
        @JvmStatic
        fun of(soundCategory: String, ringTone: String, volume: Double?): Sound =
            Sound(SoundCategory.from(soundCategory), RingTone.from(ringTone), volume!!)
    }
}
